### DISCUSSION TITLE
[A-L] (5 marzo 2024) Esercizio 2

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#3] [Esercitazioni A-L]
Un grafo $G = (G, E)$ non diretto dice bipartito se l’insieme dei vertici $V$ può essere partizionato in due insiemi disgiunti $U$ e $W$ tali che: 
1. $U \cap W = \varnothing$
2. $U \cup W = V$
3. ogni arco di $G$ è incidente ad un vertice di $U$ e ad vertice di $W$

È noto che un grafo $G$ è **bipartito** se e solo se $G$ non ha cicli di lunghezza dispari. Modificare l’algoritmo della ricerca in profondità in maniera da poter controllare se un grafo non diretto è bipartito o no, e in caso fornire una bipartizione; è possibile fare questa modifica in modo che il controllo avvenga in $\theta(|V | + |E|)$? 

Domanda bonus: nel caso in cui $G$ non è bipartito, come deve essere ulteriormente modificato l’algoritmo per ritornare un ciclo dispari di $G$?

**Comment by CasuFrost** [comment]
Assegniamo ad ogni nodo un valore booleano (che chiameremo _parità_) nel seguente modo : 
Si parte da un nodo $x$, e gli si assegna il valore $true$, si procede poi nella ricerca DFS, ed ogni volta che si passa in un nuovo nodo, 
gli si cambia il valore con l'inverso di quello precedente.

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/97389762/bf6a5c87-38aa-4391-aabe-e2f640d8ca87)

Se alla fine della ricerca vi saranno 2 nodi adiacenti con lo stesso valore, allora il grafo **non** è bipartito. 

Se G non è connesso, questa operazione va eseguita per ogni componente di G. 

Sia C l'insieme di tutte le componenti di G (se connesso, H conterrà solamente un elemento) :
```
for each g in C {
   empySet = l'insieme vuoto
   x = random node in g
   n = |V(g)|
   parità : Int = 1
   Vis : Int[n] = [0,0...,0]    //array lungo n inizializzato con tutti 0
   Part : Int[n] = [0,0...,0]    //array lungo n inizializzato con tutti 0
   Part[x]=parità
   S : stack
   S.push(x)
   while(S != emptySet){
      y=S.top()
      if(Vis[y.adiacenti[0]]==0){    //il primo vicino di y
         parità*=1
         z=y.adiacenti[0]
         Vis[z]=1
         Part[z]=parità
         S.push[z]
         y.adiacenti.remove(0) //rimuovo z dagli adiacenti di y
      }
      else{ 
         y.adiacenti.remove(0) 
      }
      if (y.adiacenti == empySet ) {
         S.pop()
      }
   }
   for each (u,v) in E(g) {
      if (Part[u]==Part[v]){
         return false;
      }
   }
}
return true;
```

**Comment by IonutCicio** [comment]
Soluzione momentanea che non ritorna il ciclo se il grafo non è bipartito, anche su grafi non connessi; Alcuni dettagli che potrebbero creare confusione:

- `mut` indica che una variabile è mutabile _(di default le variabili non possono essere modificate una volta assegnate in Rust)_
- `&` indica che sto passando la variabile per riferimento
- `true` alla fine di una funzione è sinonimo di `return true;`
- `vec![false; graph.len()]` crea un vettore di lunghezza `graph.len()` (ovvero $|V(G)|$) inizializzato a `false`

```rust
fn is_bipartite(graph: &[Vec<usize>]) -> bool {
    let mut visited = vec![false; graph.len()];
    let mut partitions = vec![false; graph.len()];

    for x in 0..graph.len() {
        if !visited[x] && !dfs_bipartite(graph, x, &mut visited, &mut partitions, true) {
            return false;
        }
    }

    true
}

fn dfs_bipartite(
    graph: &[Vec<usize>],
    x: usize,
    visited: &mut Vec<bool>,
    partitions: &mut Vec<bool>,
    partition: bool,
) -> bool {
    visited[x] = true;
    partitions[x] = partition;

    for &y in &graph[x] {
        if !visited[y] {
            if !dfs_bipartite(graph, y, visited, partitions, !partition) {
                return false;
            }
        } else if partitions[y] == partition {
            return false;
        }
    }

    true
}
```

**Comment by LeoCosta02** [comment]
![Screenshot (5)](https://github.com/sapienzastudentsnetwork/progalg2324/assets/128373356/6155a0a1-4ffd-4f99-8bd2-0869e39488d4)

**Comment by cornflexxx** [comment]
### Idea algoritmo:

Prendiamo $u_1$ , un qualunque vertice di $G$, cominciamo  allora la visita al grafo tramite $DFS$ a partire da $u_1$. Precedentemente abbiamo creato  due set, $U$ e $W$ , in $U$ per iniziare mettiamo $u_1$ , successivamente continuiamo normalmente la nostra visita, ma il secondo nodo che visitiamo , $w_1$, dovrà essere messo dentro $W$ questo perchè affinchè il grafo sia bipartito non possono esistere collegamenti diretti del tipo $u_i\rightarrow u_j$ , ma solo collegamenti che passano prima per $w_k$. Quindi ogni volta che cambiamo vertice durante la visita esso verra messo nel set "opposto" a quello del precedente: 
in breve ogni elemento presente nella lista di adiacenza di un certo $u_i$ farà parte di $W$. 

Durante la nostra visita nel momento esatto in cui incrociamo  un vertice già visitato dovremo controllare se questo fa parte dello stesso insieme di cui fa parte il vertice attualmente sulla cima dello stack. Nel caso in cui ciò sia vero ci troveremo di fronte ad un collegamento del tipo $u_i \rightarrow u_j$ o $w_k \rightarrow w_x$ questo ci garantisce che sicuaremente il grafo non è bipartito ( in precedenza avevamo detto che un grafo bipartito non presenta mai questo tipo di collegamento).

### Ritronare il ciclo dispari:
Nel caso in cui avessimo trovato che il grafo non è bipartito ci basterà svuotare lo stack fino al vertice che ha un' arco del tipo citato  sopra, in questo modo costruiamo un ciclo dispari (anche se probabilmente non è il più piccolo).

Essendo solo due le partizioni risulta più efficente la soluzione che utilizza al posto dei due insiemi degli attributi booleani da dare ai singoli vertici( di @CasuFrost ).

Nel caso in cui il grafo sia l'unione di diverse componenti possiamo  applicare l'algoritmo alle singole componenti.

**Comment by daromacarlo** [comment]
![Immagine 2024-03-16 164558](https://github.com/sapienzastudentsnetwork/progalg2324/assets/135441131/67eea6e1-96e3-4243-bc38-464d252898c9)
![Immagine 2024-03-16 164638](https://github.com/sapienzastudentsnetwork/progalg2324/assets/135441131/1c5f79d0-f2f6-467d-8d79-aa36e10506fc)

**Comment by daromacarlo** [comment]
5 marzo 2024 - esercizio 2 #3
```C
def DFS_biColorazione(G, color, x){
	C = 0
	stack = []

	color[x] = C
	cambia(C)							//funzione che fa diventare C = 1 se C == 0 e viceversa. O(1)

	stack.push(x)
	while stack is not empty{
		v = stack.pop()
		for w in v.adiacenti(){
			stack.push(w)
			if color[w] == -1{
				color[w] = C
			}
			elif color[w] == color[v]{
				return false
			}
		}
		cambia(C)					   //funzione che fa diventare C = 1 se C == 0 e viceversa. O(1)
	return true
}

def bipartito_test(G){
	color = [-1,...,-1]
    for v in V(G){
	 	if color[v] == -1{
	 		esito = DFS_biColorazione(G, color, x)
	 	}
	 	if esito == false{
	 		return false
	 	}
	}
	return true
}
``` 
-----------------------------------------------------------------
``` C
// versione ricorsiva di DFS_bicolorazione

def DFS_bicolorazione_rec(G, color, C, x)
	color[x] = C
	for w in x.adiacenti(){
		if color[w] == -1{
			DFS_bicolorazione_rec(G, color, cambia(C), w)
		}
		elif color[w] == color[x]{
			return false
		}
	}

```
